src/portable/st/synopsys/dcd_synopsys.c

```diff
@@ -677,17 +660,37 @@ static void handle_epin_ints(USB_OTG_DeviceTypeDef * dev, USB_OTG_INEndpointType
       }
 
       // XFER FIFO empty
-      if ( in_ep[n].DIEPINT & USB_OTG_DIEPINT_TXFE )
+      if ( (in_ep[n].DIEPINT & USB_OTG_DIEPINT_TXFE) && (dev->DIEPEMPMSK & (1 << n)) )
       {
         // DIEPINT's TXFE bit is read-only, software cannot clear it.
         // It will only be cleared by hardware when written bytes is more than
         // - 64 bytes or
         // - Half of TX FIFO size (configured by DIEPTXF)
 
-        transmit_packet(xfer, &in_ep[n], n);
+      	// Packets to be processed
+      	uint16_t tx_packet_amount = (in_ep[n].DIEPTSIZ & USB_OTG_DIEPTSIZ_PKTCNT_Msk) >> USB_OTG_DIEPTSIZ_PKTCNT_Pos;
+
+      	// Process every single packet (only whole packets can be written to fifo)
+      	for(uint16_t i = 0; i < tx_packet_amount; i++){
+      		// amount of bytes EP still needs to transfer
+      		uint16_t tx_remaining = (in_ep[n].DIEPTSIZ & USB_OTG_DIEPTSIZ_XFRSIZ_Msk) >> USB_OTG_DIEPTSIZ_XFRSIZ_Pos;
+      		// Packet can not be larger than ep max size
+      		uint16_t packet_size = (tx_remaining > xfer->max_size) ? xfer->max_size : tx_remaining;
+
+      		// It's only possible to write full packets into FIFO. Therefore DTXFSTS register of current
+      		// EP has to be checked if the buffer can take another WHOLE packet
+      		if(packet_size > ((in_ep[n].DTXFSTS & USB_OTG_DTXFSTS_INEPTFSAV_Msk) << 2)){
+      			break;
+      		}
+
+      		xfer->queued_len = xfer->total_len - tx_remaining;
+
+      		// Push packet to Tx-FIFO
+					transmit_fifo_packet(n, (xfer->buffer + xfer->queued_len), packet_size);
+      	}
 
         // Turn off TXFE if all bytes are written.
-        if (xfer->queued_len == xfer->total_len)
+        if (((in_ep[n].DIEPTSIZ & USB_OTG_DIEPTSIZ_XFRSIZ_Msk) >> USB_OTG_DIEPTSIZ_XFRSIZ_Pos) == 0)
```

Comment:
@cr1901  Yeah, you are right `xfer->queued_len` won't be equal to total_len since the last interrupt is the XFRC one. Though I think we could infer the `queue_len` using the DIEPTSIZ and can actually remove the usage of `queued_len` in the future as well. 

@duempel, though looking at the manual, I have another question. The specs said the **XFRSIZ is decreased when a packet is written to TxFIFO**. I wonder if it also include short packet, I mean the core could detect the short packet and decrease it when we check of this right ? at worst we got into ISR once more time like previously ?  

![image](https://user-images.githubusercontent.com/249515/80856551-54345700-8c75-11ea-9643-ee46ec972bd7.png)